### PR TITLE
Define KOKKOS_COMPILER_INTEL_LLVM and only define at most one KOKKOS_COMPILER* macro

### PR DESCRIPTION
**Pull request by masterleinad**
As discussed this pull request defines a new compiler macro that allows distinguishing between the Intel Classic Compiler (icpc) and the Intel oneAPI Compiler (icpx). I went through the places that use `KOKKOS_COMPILER_INTEL_LLVM` and replaced/added/omitted where I saw fit.

Also, this pull request makes sure that only at most one compiler macro `KOKKOS_COMPILER*` is defined (apart from `KOKKOS_COMPILER_NVCC` which can appear additionally). I'm happy to separate this change from this pull request if preferred.

### CODE REVIEW

File: core/src/Kokkos_Macros.hpp [hunk 1]
```diff
@@ -148,12 +149,14 @@
 #define KOKKOS_COMPILER_APPLECC __APPLE_CC__
 #endif
 
-#if defined(__clang__) && !defined(KOKKOS_COMPILER_INTEL)
+#if defined(__clang__) && !defined(KOKKOS_COMPILER_INTEL) && \
+    !defined(KOKKOS_COMPILER_INTEL_LLVM)
```
Comment (PhilMiller) [inline core/src/Kokkos_Macros.hpp:153]:
```suggestion
#if defined(__clang__) && !defined(KOKKOS_COMPILER_INTEL_LLVM)
```
I don't think there's a case where `KOKKOS_COMPILER_INTEL` is relevant to this, but the `LLVM` macro isn't defined.

File: core/src/Kokkos_Macros.hpp [hunk 2]
```diff
@@ -130,7 +131,7 @@
 #if defined(__INTEL_COMPILER)
 #define KOKKOS_COMPILER_INTEL __INTEL_COMPILER
 #elif defined(__INTEL_LLVM_COMPILER)
-#define KOKKOS_COMPILER_INTEL __INTEL_LLVM_COMPILER
+#define KOKKOS_COMPILER_INTEL_LLVM __INTEL_LLVM_COMPILER
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/Kokkos_Macros.hpp:134]: No, https://www.intel.com/content/www/us/en/developer/articles/technical/use-predefined-macros-for-specific-code-for-intel-dpcpp-compiler-intel-cpp-compiler-intel-cpp-compiler-classic.html.

File: core/src/Kokkos_Macros.hpp [hunk 3]
```diff
@@ -182,19 +186,16 @@
 //----------------------------------------------------------------------------
 // Intel compiler macros
 
-#if defined(KOKKOS_COMPILER_INTEL)
-// FIXME_ICPX
-#if !defined(__INTEL_LLVM_COMPILER)
+// FIXME_INTEL_LLVM
+#if defined(KOKKOS_COMPILER_INTEL_LLVM)
```
Comment (PhilMiller) [inline core/src/Kokkos_Macros.hpp:189]: The sense of the test got inverted here, I think? It had been classic `icpc` only. I think the `FIXME` comment was to see whether we should be enabling these pragmas for that as well.

Comment (masterleinad) [reply core/src/Kokkos_Macros.hpp]: Fixed.

File: core/src/Kokkos_Macros.hpp [hunk 4]
```diff
@@ -182,19 +185,16 @@
 //----------------------------------------------------------------------------
 // Intel compiler macros
 
+// FIXME_INTEL_LLVM
```
Comment (PhilMiller) [inline core/src/Kokkos_Macros.hpp:188]:
```suggestion
// FIXME_INTEL_LLVM: Figure out whether these pragmas should get set similarly to the classic Intel compiler
```

File: core/src/Kokkos_Macros.hpp [hunk 5]
```diff
@@ -148,12 +149,13 @@
...
 
-#if defined(__clang__) && !defined(KOKKOS_COMPILER_INTEL)
+#if defined(__clang__) && !defined(KOKKOS_COMPILER_INTEL_LLVM)
 #define KOKKOS_COMPILER_CLANG \
   __clang_major__ * 100 + __clang_minor__ * 10 + __clang_patchlevel__
 #endif
 
-#if !defined(__clang__) && !defined(KOKKOS_COMPILER_INTEL) && defined(__GNUC__)
+#if !defined(__clang__) && !defined(KOKKOS_COMPILER_INTEL) && \
+    !defined(KOKKOS_COMPILER_INTEL_LLVM) && defined(__GNUC__)
```
Comment (PhilMiller) [inline core/src/Kokkos_Macros.hpp:158]: I don't we need to check `INTEL_LLVM` here, since `__clang__` will be defined in that case.

Comment (PhilMiller) [reply core/src/Kokkos_Macros.hpp:158]: I.e. this line can just be reverted.

File: core/src/Kokkos_Macros.hpp [hunk 6]
```diff
@@ -130,45 +131,39 @@
...
 #define KOKKOS_COMPILER_INTEL __INTEL_COMPILER
 #elif defined(__INTEL_LLVM_COMPILER)
-#define KOKKOS_COMPILER_INTEL __INTEL_LLVM_COMPILER
+#define KOKKOS_COMPILER_INTEL_LLVM __INTEL_LLVM_COMPILER
 #elif defined(__ICC)
 // Old define
 #define KOKKOS_COMPILER_INTEL __ICC
 #elif defined(__ECC)
 // Very old define
 #define KOKKOS_COMPILER_INTEL __ECC
```
Comment (PhilMiller) [inline core/src/Kokkos_Macros.hpp:140]: These are probably no longer relevant, since we bumped the minimum supported version to one that defines `__INTEL_COMPILER`

File: core/src/Kokkos_Macros.hpp [hunk 7]
```diff
@@ -130,45 +131,39 @@
...
-#endif
 
+#elif defined(_CRAYC)
 // CRAY compiler for host code
-#if defined(_CRAYC)
 #define KOKKOS_COMPILER_CRAYC _CRAYC
-#endif
 
-#if defined(__APPLE_CC__)
+#elif defined(__APPLE_CC__)
```
Comment (PhilMiller) [inline core/src/Kokkos_Macros.hpp:146]: Since you're now making these definitions strictly mutually exclusive, please check that any places that were testing `KOKKOS_COMPILER_CLANG` are suitably adapted to check `defined(KOKKOS_COMPILER_CLANG) || defined(KOKKOS_COMPILER_APPLECC)`

File: core/src/Kokkos_Macros.hpp [hunk 8]
```diff
@@ -130,45 +131,39 @@
...
-#endif
 
-#if defined(__clang__) && !defined(KOKKOS_COMPILER_INTEL)
+#elif defined(__clang__)
 #define KOKKOS_COMPILER_CLANG \
   __clang_major__ * 100 + __clang_minor__ * 10 + __clang_patchlevel__
-#endif
 
-#if !defined(__clang__) && !defined(KOKKOS_COMPILER_INTEL) && defined(__GNUC__)
+#elif defined(__GNUC__)
```
Comment (PhilMiller) [inline core/src/Kokkos_Macros.hpp:153]:
```suggestion
// Check this last because many compilers (at least Clang variants and Intel classic) define `__GNUC__` for compatibility
#elif defined(__GNUC__)
```

File: core/src/Kokkos_Macros.hpp [hunk 9]
```diff
@@ -130,45 +131,39 @@
...
 #define KOKKOS_COMPILER_CRAYC _CRAYC
-#endif
 
-#if defined(__APPLE_CC__)
+#elif defined(__APPLE_CC__)
 #define KOKKOS_COMPILER_APPLECC __APPLE_CC__
-#endif
 
-#if defined(__clang__) && !defined(KOKKOS_COMPILER_INTEL)
+#elif defined(__clang__)
```
Comment (PhilMiller) [inline core/src/Kokkos_Macros.hpp:149]:
```suggestion
// Check this after the Clang-based proprietary compilers which will also define __clang__
#elif defined(__clang__)
```

File: core/src/Kokkos_Macros.hpp [hunk 10]
```diff
@@ -130,45 +131,39 @@
...
 #define KOKKOS_COMPILER_CLANG \
   __clang_major__ * 100 + __clang_minor__ * 10 + __clang_patchlevel__
-#endif
 
-#if !defined(__clang__) && !defined(KOKKOS_COMPILER_INTEL) && defined(__GNUC__)
+#elif defined(__GNUC__)
 #define KOKKOS_COMPILER_GNU \
   __GNUC__ * 100 + __GNUC_MINOR__ * 10 + __GNUC_PATCHLEVEL__
 
 #if (530 > KOKKOS_COMPILER_GNU)
```
Comment (PhilMiller) [inline core/src/Kokkos_Macros.hpp:157]: I just noticed - this should be 820 now. I think that's perhaps already enforced in CMake, so it might be redundant. It might not be getting checked with Makefiles, though.

Comment (PhilMiller) [reply core/src/Kokkos_Macros.hpp:157]: https://github.com/kokkos/kokkos/issues/5285

File: core/src/Kokkos_Macros.hpp [hunk 11]
```diff
@@ -130,45 +131,39 @@
...
 
-#if defined(__NVCOMPILER)
+#elif defined(__NVCOMPILER)
 #define KOKKOS_COMPILER_NVHPC                                 \
   __NVCOMPILER_MAJOR__ * 10000 + __NVCOMPILER_MINOR__ * 100 + \
       __NVCOMPILER_PATCHLEVEL__
-#endif
 
-#if defined(_MSC_VER) && !defined(KOKKOS_COMPILER_INTEL)
+#elif defined(_MSC_VER)
```
Comment (PhilMiller) [inline core/src/Kokkos_Macros.hpp:166]:
```suggestion
// Check this after Intel because that defines _MSC_VER for compatibility
#elif defined(_MSC_VER)
```

Does Clang on Windows also do that, like it defines `__GNUC__` on platforms where GCC was historically the default?

File: core/src/Kokkos_Macros.hpp (hunk 7 above)
Comment (masterleinad) [reply core/src/Kokkos_Macros.hpp:146]: I did now. We only really check for `Clang` when `CUDA` or `OpenMPTarget` is enabled and `AppleClang`  wouldn't work in these cases anyway. There is of course also the check for builtins that need some love anyway.

File: core/src/Kokkos_Macros.hpp (hunk 6 above)
Comment (masterleinad) [reply core/src/Kokkos_Macros.hpp:140]: Removed.

File: core/src/Kokkos_Macros.hpp (hunk 11 above)
Comment (masterleinad) [reply core/src/Kokkos_Macros.hpp:166]: Yes.

File: example/build_cmake_in_tree/cmake_example.cpp [hunk 12]
```diff
@@ -14,47 +14,8 @@
...
-  }
-
-  count_time = timer.seconds();
-  printf("Sequential: %ld    %10.6f\n", seq_count, count_time);
-
-  Kokkos::finalize();
-
-  return (count == seq_count) ? 0 : -1;
-}
+int main(int argc, char* argv[]) { bla(); }
```
Comment (PhilMiller) [inline example/build_cmake_in_tree/cmake_example.cpp:21]: This file's changes look like they're addressing Andrey's request for an abort header, swept up in commit -a

**Review by PhilMiller — COMMENTED** [review]
Things generally look good now. I haven't grepped for all of the uses of the affected macros myself to confirm that there are no missing updates.

File: example/build_cmake_in_tree/cmake_example.cpp (hunk 12 above)
Comment (masterleinad) [reply example/build_cmake_in_tree/cmake_example.cpp]: Fixed.

**Comment by PhilMiller** [comment]
I've checked over all of the uses of `KOKKOS_COMPILER_INTEL` in Kokkos and Kokkos Kernels. Everything else in Kokkos is good, with maybe some `#if defined(KOKKOS_COMPILER_INTEL` gaining `|| defined(KOKKOS_COMPILER_INTEL_LLVM)`, but none of them needing to be substituted or whatever. The two uses in Kernels will end up being simplified, because they're currently `defined(INTEL) && !defined(clang)`.

There's one non-Kokkos usage of the macro in Trilinos to work around an Intel ICE [in the Kokkos 3.5 PR](https://github.com/trilinos/Trilinos/pull/9836/commits)
https://github.com/trilinos/Trilinos/blame/master/packages/intrepid2/src/Shared/Intrepid2_Data.hpp#L1688
I'll just post a PR to try removing that, and see if it blows up in testing.

**Review by PhilMiller — APPROVED** [review]

File: core/src/Kokkos_BitManipulation.hpp [hunk 13]
```diff
@@ -183,6 +183,7 @@ rotr(T x, int s) noexcept {
 
 namespace Kokkos::Impl {
 
+// FIXME Try to use intrinsics for more compilers
```
Comment (dalg24) [inline core/src/Kokkos_BitManipulation.hpp:186]: Drive-by change that does not belong to this PR

File: core/src/Kokkos_Macros.hpp [hunk 14]
```diff
@@ -129,46 +130,41 @@
...
 #define KOKKOS_COMPILER_CRAYC _CRAYC
-#endif
 
-#if defined(__APPLE_CC__)
+#elif defined(__APPLE_CC__)
 #define KOKKOS_COMPILER_APPLECC __APPLE_CC__
-#endif
 
-#if defined(__clang__) && !defined(KOKKOS_COMPILER_INTEL)
+#elif defined(__clang__)
```
Comment (dalg24) [inline core/src/Kokkos_Macros.hpp:144]: Your PR is titled "Define KOKKOS_COMPILER_INTEL_LLVM" and this change implies for instance that now an AppleClang build still has `KOKKOS_COMPILER_APPLECC` but not `KOKKOS_COMPILER_CLANG ` any more.
Although I don't necessary object with the direction, this is clearly out of the advertised scope.

**Review by dalg24 — CHANGES_REQUESTED** [review]

**Comment by PhilMiller** [comment]
Would be good to fix the commit message to reflect the full scope of the PR.

File: core/src/Kokkos_BitManipulation.hpp (hunk 13 above)
Comment (dalg24) [reply core/src/Kokkos_BitManipulation.hpp:226]: Ping

**Review by dalg24 — DISMISSED** [review]
Please add a `unit_tests/TestCompilerMacros.cpp` and check that one and only one macro is being defined.

File: core/src/Kokkos_Macros.hpp [hunk 15]
```diff
@@ -176,19 +172,17 @@
 //----------------------------------------------------------------------------
 // Intel compiler macros
 
+// FIXME_INTEL_LLVM: Figure out whether these pragmas should get set similarly
+// to the classic Intel compiler
```
Comment (crtrott) [inline core/src/Kokkos_Macros.hpp:176]: yes

**Review by crtrott — CHANGES_REQUESTED** [review]
I think those other macros like ivdep should be defined for Intel LLVM

File: core/unit_test/TestCompilerMacros.hpp [hunk 16]
```diff
@@ -16,6 +16,70 @@
 
 #include <Kokkos_Core.hpp>
 
+#ifdef KOKKOS_COMPILER_INTEL
```
Comment (dalg24) [inline core/unit_test/TestCompilerMacros.hpp:19]: I was more thinking something a la
https://github.com/kokkos/kokkos/blob/aa1f48f3172069f212ed2c97b74b786999b26f17/core/src/Kokkos_Macros.hpp#L408-L418

File: core/unit_test/TestCompilerMacros.hpp [hunk 17]
```diff
@@ -16,6 +16,17 @@
 
 #include <Kokkos_Core.hpp>
 
+#if 1 < ((defined(KOKKOS_COMPILER_INTEL) ? 1 : 0) +      \
```
Comment (dalg24) [inline core/unit_test/TestCompilerMacros.hpp:19]: not greater than 1, equal to 1

File: core/unit_test/TestCompilerMacros.hpp [hunk 18]
```diff
@@ -51,7 +62,7 @@ struct AddFunctor {
 #pragma vector always
 #endif
 #ifdef KOKKOS_ENABLE_PRAGMA_LOOPCOUNT
-#pragma loop count(128)
+#pragma loop_count(128)
```
Comment (dalg24) [inline core/unit_test/TestCompilerMacros.hpp:65]: uggggg

File: .github/workflows/continuous-integration-workflow.yml [hunk 19]
```diff
@@ -29,12 +29,12 @@ jobs:
             backend: 'OPENMP'
           - distro: 'fedora:intel'
             cxx: 'icpx'
-            cxx_extra_flags: '-fp-model=precise'
+            cxx_extra_flags: '-fp-model=precise -Wno-pass-failed'
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply .github/workflows/continuous-integration-workflow.yml:32]: I get the warning that some loops can not be optimized as requested when enabling the pragmas for OpenMP with `icpx`.

[the remark this answers is not in the document]
Comment (dalg24) [reply .github/workflows/continuous-integration-workflow.yml:32]: These warnings were not there before?

[the remark this answers is not in the document]
Comment (masterleinad) [reply .github/workflows/continuous-integration-workflow.yml:32]: No, we are only now enabling a bunch of pragmas (usable in host code; depending on the compiler version).

**Comment by masterleinad** [comment]
Only `gcc-8.4.0` is failing with
```
KokkosContainers_UnitTest_OpenMP (Child killed)
```

**Review by crtrott — APPROVED** [review]

File: core/src/Kokkos_Macros.hpp [hunk 20]
```diff
@@ -123,46 +124,41 @@
...
 #elif defined(__INTEL_LLVM_COMPILER)
-#define KOKKOS_COMPILER_INTEL __INTEL_LLVM_COMPILER
-#elif defined(__ICC)
-// Old define
-#define KOKKOS_COMPILER_INTEL __ICC
-#elif defined(__ECC)
-// Very old define
-#define KOKKOS_COMPILER_INTEL __ECC
-#endif
+#define KOKKOS_COMPILER_INTEL_LLVM __INTEL_LLVM_COMPILER
```
Comment (rgayatri23) [inline core/src/Kokkos_Macros.hpp:129]: Thanks. I will follow this concept for the Cray one too in #5889

**Review by rgayatri23 — APPROVED** [review]

**Comment by dalg24** [comment]
Ignoring GCC-8.4.0 failure which is clearly unrelated (running out of memory)
This was resolved elsewhere
